File: clients/Ruby/model/move_action.rb
require_relative 'vec2_int'
class MoveAction
    attr_accessor :target
    attr_accessor :find_closest_position
    attr_accessor :break_through
    def initialize(target, find_closest_position, break_through)
        @target = target
        @find_closest_position = find_closest_position
        @break_through = break_through
    end
    def self.read_from(stream)
        target = Vec2Int.read_from(stream)
        find_closest_position = stream.read_bool()
        break_through = stream.read_bool()
        MoveAction.new(target, find_closest_position, break_through)
    end
    def write_to(stream)
        @target.write_to(stream)
        stream.write_bool(@find_closest_position)
        stream.write_bool(@break_through)
    end
end

File: clients/TypeScript/src/model/MoveAction.ts
import {Stream} from './Stream';
import {Vec2Int} from './Vec2Int';

export class MoveAction {
    constructor(
        public target: Vec2Int,
        public findClosestPosition: boolean,
        public breakThrough: boolean,
    ) {}

    static async readFrom(stream: Stream) {
        const target = await Vec2Int.readFrom(stream);
        const findClosestPosition = await stream.readBool();
        const breakThrough = await stream.readBool();

        return new MoveAction(target, findClosestPosition, breakThrough);
    }
    async writeTo(stream: Stream) {
        await this.target.writeTo(stream);
        await stream.writeBool(this.findClosestPosition);
        await stream.writeBool(this.breakThrough);
    }
}
